internal/databases/postgres/catchup_send_recieve_handler.go

```diff
@@ -0,0 +1,289 @@
+	"os"
+	"path"
+	"path/filepath"
+)
+
+func HandleCatchupSend(pgDataDirectory string, destination string) {
+	pgDataDirectory = utility.ResolveSymlink(pgDataDirectory)
+	tracelog.InfoLogger.Printf("Sending %v to %v\n", pgDataDirectory, destination)
+	info, runner, err := GetPgServerInfo(true)
+	if info.systemIdentifier == nil {
+		tracelog.ErrorLogger.Fatal("Our system lacks System Identifier, cannot proceed")
+	}
+	tracelog.ErrorLogger.FatalOnError(err)
+	dial, err := net.Dial("tcp", destination)
+	tracelog.ErrorLogger.FatalOnError(err)
+	decoder := gob.NewDecoder(dial)
+	var control PgControlData
+	err = decoder.Decode(&control)
+	tracelog.ErrorLogger.FatalOnError(err)
+	tracelog.InfoLogger.Printf("Destination control file %v", control)
+	tracelog.InfoLogger.Printf("Our system id %v", *info.systemIdentifier)
+	if *info.systemIdentifier != control.SystemIdentifier {
+		tracelog.ErrorLogger.Fatal("System identifiers do not match")
+	}
+	if control.CurrentTimeline != info.Timeline {
+		tracelog.ErrorLogger.Fatalf("Destination is on timeline %v, but we are on %v", control.CurrentTimeline, info.Timeline)
+	}
+	var fileList internal.BackupFileList
+	err = decoder.Decode(&fileList)
+	tracelog.ErrorLogger.FatalOnError(err)
+	tracelog.InfoLogger.Printf("Received file list of %v failes", len(fileList))
+	_, lsnStr, _, err := runner.StartBackup("")
+	tracelog.ErrorLogger.FatalOnError(err)
+	lsn, err := ParseLSN(lsnStr)
+	tracelog.ErrorLogger.FatalOnError(err)
+	if lsn <= control.Checkpoint {
+		tracelog.ErrorLogger.Fatalf("Catchup destination is already ahead (our LSN %v, destination LSN %v).", lsn, control.Checkpoint)
+	}
+
+	encoder := gob.NewEncoder(dial)
+
+	label, offsetMap, _, err := runner.StopBackup()
+	tracelog.ErrorLogger.FatalOnError(err)
+
+	sendFileCommands(encoder, pgDataDirectory, fileList, control.Checkpoint)
+
+	err = encoder.Encode(CatchupCommandDto{BinaryContents: []byte(label), FileName: BackupLabelFilename, IsBinContents: true})
+	tracelog.ErrorLogger.FatalOnError(err)
+	err = encoder.Encode(CatchupCommandDto{BinaryContents: []byte(offsetMap), FileName: TablespaceMapFilename, IsBinContents: true})
+	tracelog.ErrorLogger.FatalOnError(err)
+	ourPgControl, err := os.ReadFile(path.Join(pgDataDirectory, PgControlPath))
+	tracelog.ErrorLogger.FatalOnError(err)
+	err = encoder.Encode(CatchupCommandDto{BinaryContents: ourPgControl, FileName: utility.SanitizePath(PgControlPath), IsBinContents: true})
+	tracelog.ErrorLogger.FatalOnError(err)
+
+	err = encoder.Encode(CatchupCommandDto{IsDone: true})
+	tracelog.ErrorLogger.FatalOnError(err)
+	tracelog.InfoLogger.Printf("Send done")
+}
+
+func sendFileCommands(encoder *gob.Encoder, directory string, list internal.BackupFileList, checkpoint LSN) {
+	extendExcludedFiles()
+	err := filepath.Walk(directory, func(path string, info fs.FileInfo, err error) error {
+		if err != nil {
+			if os.IsNotExist(err) {
+				tracelog.WarningLogger.Println(path, " deleted during filepath walk")
+				return nil
```

Comment:
Is it possible that PostgreSQL removes file during `catchup-send`? (it looks like you are taking database locks only for metadata reading)

If files can be removed:
 * should we remove it on replica? (like `rsync --delete`)
 * could we observe deleted file in line 114/121?

Reply:
Yup, we should remove files that are not present on Standby. I'm planing to add commands for this now.

Reply:
>could we observe deleted file in line 114/121?
Yes we can. I'll fix that.